stdlib/public/runtime/GenericMetadataBuilder.cpp

```diff
@@ -391,6 +392,163 @@ static bool equalVWTs(const ValueWitnessTable *a, const ValueWitnessTable *b) {
   return false;
 }
 
+bool swift::compareGenericMetadata(const Metadata *original,
+                                   const Metadata *newMetadata) {
+  if (original == newMetadata)
+    return true;
+
+  bool equal = true;
+
+  if (original->getKind() != newMetadata->getKind()) {
+    validationLog(true, "Kinds do not match");
+    equal = false;
+  } else {
+    auto originalDescriptor = original->getTypeContextDescriptor();
+    auto newDescriptor = newMetadata->getTypeContextDescriptor();
+
+    if (originalDescriptor != newDescriptor) {
+      validationLog(true, "Descriptors do not match");
+      equal = false;
+    } else if (!originalDescriptor->isGeneric()) {
+      validationLog(true,
+                    "Descriptor is not generic and pointers are not identical");
+    } else {
+      auto origVWT = asFullMetadata(original)->ValueWitnesses;
+      auto newVWT = asFullMetadata(newMetadata)->ValueWitnesses;
+
+      if (!equalVWTs(origVWT, newVWT)) {
+        validationLog(true, "VWTs do not match");
+        equal = false;
+      }
+
+      size_t originalSize = 0;
+      size_t newSize = 0;
+
+      // Start with a minimal static size;
+      switch (original->getKind()) {
+      case MetadataKind::Class:
+        originalSize = newSize = sizeof(ClassMetadata);
+        break;
+      case MetadataKind::Struct:
+        originalSize = newSize = sizeof(StructMetadata);
```

Comment:
You're right, I think I was attempting to handle non-generic metadata as well so that the subsequent size computation with `genericArgumentsEnd` might not happen. Then I decided that distinct non-generic metadata is always non-equal.